src/compiler/parser.ts

```diff
@@ -3445,6 +3445,14 @@ namespace Parser {
                 // out a comma so we give a good error message.
                 parseExpected(SyntaxKind.CommaToken, getExpectedCommaDiagnostic(kind));
 
+                // since we didn't get a comma it might be better finish the list if the next token is {
+                // this allows us to avoid parsing objects as binding patterns
+                // as that might lead to incorectly finishing class declarations when parsing the pattern fails
+                // and leftover } are used for finishing the current method and the class declaration
+                if (!(contextFlags & NodeFlags.JsonFile) && token() === SyntaxKind.OpenBraceToken) {

```

Comment:
This might be an overly simple check - any improvements to this are welcome.

Reply:
At a glance, it _feels_ like there should be somewhere else to stick this or some other condition that uses some context; this function is like the main loop that parses out list, but this is a specific case that really depends on the grammar.

I'd have to think about this harder, though.